### PR TITLE
Fix tpm2_checkquote when run on a big endian system

### PR DESCRIPTION
**Pull request by stefanberger**
Various PCR related structures read by tpm2_checkquote are written
in little endian format, which needs to be taken into account when
run on a big endian system. Although little endian is unusual for
TPM, it is better to leave these data structures in little endian
format to not disrupt existing systems.

Signed-off-by: Stefan Berger <stefanb@linux.ibm.com>

### CODE REVIEW

**Comment by stefanberger** [comment]
I added a patch now that enforces that `tpm2_quote` writes its data in little endian format, which is not invasive for existing x86 systems. I have tested this now with `tpm2_quote` running on both big and little endian system and the `tpm2_checkquote` running on a big endian system.

File: lib/pcr.c [hunk 1]
```diff
@@ -177,6 +178,59 @@ static bool pcr_unset_pcr_sections(TPML_PCR_SELECTION *s) {
...
 }
 
+bool pcr_print_pcr_struct_le(TPML_PCR_SELECTION *pcr_select, tpm2_pcrs *pcrs) {
+
+    UINT32 vi = 0, di = 0, i;
+    bool result = true;
+
+    tpm2_tool_output("pcrs:\n");
+
+    // Loop through all PCR/hash banks
```
Comment (idesai) [inline lib/pcr.c:188]: line comments to be used at end of line.
```suggestion
    /* Loop through all PCR/hash banks */
```

File: lib/pcr.c [hunk 2]
```diff
@@ -177,6 +178,59 @@ static bool pcr_unset_pcr_sections(TPML_PCR_SELECTION *s) {
...
+    for (i = 0; i < le32toh(pcr_select->count); i++) {
+        const char *alg_name = tpm2_alg_util_algtostr(
+                le16toh(pcr_select->pcrSelections[i].hash), tpm2_alg_util_flags_hash);
+
+        tpm2_tool_output("  %s:\n", alg_name);
+
+        // Loop through all PCRs in this bank
```
Comment (idesai) [inline lib/pcr.c:195]:
```suggestion
        /* Loop through all PCRs in this bank */
```

File: lib/pcr.c [hunk 3]
```diff
@@ -177,6 +178,59 @@ static bool pcr_unset_pcr_sections(TPML_PCR_SELECTION *s) {
...
+        unsigned int pcr_id;
+        for (pcr_id = 0; pcr_id < pcr_select->pcrSelections[i].sizeofSelect * 8u;
+                pcr_id++) {
+            if (!tpm2_util_is_pcr_select_bit_set(&pcr_select->pcrSelections[i],
+                    pcr_id)) {
+                // skip non-selected banks
+                continue;
```
Comment (idesai) [inline lib/pcr.c:202]:
```suggestion
                continue; //skip non-selected banks
```

File: lib/pcr.c [hunk 4]
```diff
@@ -177,6 +178,59 @@ static bool pcr_unset_pcr_sections(TPML_PCR_SELECTION *s) {
...
+            }
+            if (vi >= le64toh(pcrs->count) || di >= le32toh(pcrs->pcr_values[vi].count)) {
+                LOG_ERR("Something wrong, trying to print but nothing more");
+                return false;
+            }
+
+            // Print out PCR ID
```
Comment (idesai) [inline lib/pcr.c:209]:
```suggestion
            /* Print out PCR ID */
```

File: lib/pcr.c [hunk 5]
```diff
@@ -177,6 +178,59 @@ static bool pcr_unset_pcr_sections(TPML_PCR_SELECTION *s) {
...
+            tpm2_tool_output("    %-2d: 0x", pcr_id);
+
+            // Print out current PCR digest value
```
Comment (idesai) [inline lib/pcr.c:212]:
```suggestion
            /* Print out current PCR digest value */
```

File: lib/systemdeps.h [hunk 6]
```diff
@@ -0,0 +1,12 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
```
Comment (idesai) [inline lib/systemdeps.h:1]: Please rename this file to tpm2_systemdeps.h

File: lib/tpm2_openssl.c [hunk 7]
```diff
@@ -343,6 +344,83 @@ bool tpm2_openssl_hash_pcr_banks(TPMI_ALG_HASH hash_alg,
     return result;
 }
 
+// show all PCR banks according to g_pcrSelection & g_pcrs->
```
Comment (idesai) [inline lib/tpm2_openssl.c:347]:
```suggestion
/* show all PCR banks according to g_pcrSelection & g_pcrs-> */
```

File: lib/tpm2_openssl.c [hunk 8]
```diff
@@ -343,6 +344,83 @@ bool tpm2_openssl_hash_pcr_banks(TPMI_ALG_HASH hash_alg,
...
+        return false;
+    }
+
+    int rc = EVP_DigestInit_ex(mdctx, md, NULL);
+    if (!rc) {
+        LOG_ERR("%s", tpm2_openssl_get_err());
+        goto out;
+    }
+
+    // Loop through all PCR/hash banks
```
Comment (idesai) [inline lib/tpm2_openssl.c:371]:
```suggestion
    /* Loop through all PCR/hash banks */
```

File: lib/tpm2_openssl.c [hunk 9]
```diff
@@ -343,6 +344,83 @@ bool tpm2_openssl_hash_pcr_banks(TPMI_ALG_HASH hash_alg,
...
+    for (i = 0; i < le32toh(pcr_select->count); i++) {
+
+        // Loop through all PCRs in this bank
```
Comment (idesai) [inline lib/tpm2_openssl.c:374]:
```suggestion
        /* Loop through all PCRs in this bank */
```

File: lib/tpm2_openssl.c [hunk 10]
```diff
@@ -343,6 +344,83 @@ bool tpm2_openssl_hash_pcr_banks(TPMI_ALG_HASH hash_alg,
...
+        unsigned int pcr_id;
+        for (pcr_id = 0; pcr_id < pcr_select->pcrSelections[i].sizeofSelect * 8u;
+                pcr_id++) {
+            if (!tpm2_util_is_pcr_select_bit_set(&pcr_select->pcrSelections[i],
+                    pcr_id)) {
+                // skip non-selected banks
+                continue;
```
Comment (idesai) [inline lib/tpm2_openssl.c:381]:
```suggestion
                continue; //skip non-selected banks
```

File: lib/tpm2_openssl.c [hunk 11]
```diff
@@ -343,6 +344,83 @@ bool tpm2_openssl_hash_pcr_banks(TPMI_ALG_HASH hash_alg,
...
+            }
+            if (vi >= le64toh(pcrs->count) || di >= le32toh(pcrs->pcr_values[vi].count)) {
+                LOG_ERR("Something wrong, trying to print but nothing more");
+                goto out;
+            }
+
+            // Update running digest (to compare with quote)
```
Comment (idesai) [inline lib/tpm2_openssl.c:388]:
```suggestion
            /* Update running digest (to compare with quote) */
```

File: lib/tpm2_openssl.c [hunk 12]
```diff
@@ -343,6 +344,83 @@ bool tpm2_openssl_hash_pcr_banks(TPMI_ALG_HASH hash_alg,
...
+            }
+
+            di = 0;
+            if (++vi < le64toh(pcrs->count)) {
+                continue;
+            }
+        }
+    }
+
+    // Finalize running digest
```
Comment (idesai) [inline lib/tpm2_openssl.c:407]:
```suggestion
    /* Finalize running digest */
```

File: tools/misc/tpm2_checkquote.c [hunk 13]
```diff
@@ -11,6 +11,7 @@
 #include "files.h"
 #include "log.h"
 #include "object.h"
+#include "systemdeps.h"
```
Comment (idesai) [inline tools/misc/tpm2_checkquote.c:14]:
```suggestion
#include "tpm2_systemdeps.h"
```

**Review by idesai — CHANGES_REQUESTED** [review]

**Comment by idesai** [comment]
- [ ] Test build with qemu-arm32v7

**Pull request merged**